src/vga/src/lib.rs

```diff
@@ -67,32 +68,15 @@ impl VgaBuffer {
             // to get the current line, we divide by the length of a line
             let current_line = (self.position as isize) / CONSOLE_COLS;
 
-            let next_line = if current_line + 1 >= CONSOLE_ROWS {
-
-                let end = CONSOLE_ROWS * CONSOLE_COLS;
-
-                for i in CONSOLE_COLS..(end) {
-                    let prev = i - CONSOLE_COLS;
-                    self.buffer[prev as usize] = self.buffer[i as usize];
-
-                }
-
-                // blank out the last row
-                for i in (end - CONSOLE_COLS)..(end) {
-                    let cell = &mut self.buffer[i as usize];
-                    *cell = VgaCell {
-                        character: ' ' as u8,
-                        color: DEFAULT_COLOR,
-                    };
-                }
-
-                CONSOLE_ROWS - 1
+            if current_line + 1 >= CONSOLE_ROWS {
+                self.scroll_up();
             } else {
-                current_line + 1
+                self.position = ((current_line + 1) * CONSOLE_COLS) as usize;
             };
-
-            self.position = (next_line * CONSOLE_COLS) as usize;
         } else {
+            if self.position >= self.buffer.len() {
```

Comment:
another `if` inside of an `else` makes me wonder if maybe there isn't more refactoring to be done here. It's late and I haven't read the full file, just the diff, so maybe not, but also makes me wonder...